Apps/Src/ReadCarCAN.c

```diff
@@ -101,65 +89,126 @@ static inline void chargingEnable(void) {
     assertOSError(OS_READ_CAN_LOC,err);
 }
 
+/**
+ * @brief Callback function for the precharge delay timer. Turns off precharge and restarts the array.
+ * 
+*/
+static void arrayRestart(void *p_tmr, void *p_arg){
+
+    if(!RegenEnable){    // If regen enable has been disabled during precharge, we don't want to turn on the main contactor immediately after
+        restartingArray = false;
+    }
+    else {
+        // Try to turn on array, but only if we can successfully turn off precharge
+        if(Contactors_Set(ARRAY_PRECHARGE, OFF, SET_CONTACTORS_NON_BLOCKING) == SUCCESS){
+
+            Contactors_Set(ARRAY_CONTACTOR, ON, SET_CONTACTORS_NON_BLOCKING);
+
+        }
+    }
+
+        Lights_Set(A_CNCTR, ON);
+
+        // done restarting the array
+        restartingArray = false;
+
+};
+
+/**
+ * @brief Disables charging and increments the trip counter
+ * when canWatchTimer hits 0 (when charge enable messages are missed)
+ * 
+*/
+void canWatchTimerCallback (void *p_tmr, void *p_arg){  //Probably needs its timer arguments
+    OS_ERR err;
+
+    // mark regen as disabled
+    RegenEnable = OFF;
+
+    //increment trip counter
+    CANWatchTripCounter++;
+    
+    // Set fault bitmaps 
+    FaultBitmap |= FAULT_READBPS;
+    OSErrLocBitmap |= OS_READ_CAN_LOC;
+
+    // If we tripped the counter too many times, enter fault state
+    if(CANWatchTripCounter > CANWATCH_TRIP_LIMIT) {
```

Comment:
Calling EnterFaultState and posting to the fault state semaphore will both do the same thing, the only difference is whether the fromThread variable is set or not. You probably want the trip counter to be in fault state and increment it there, so fault state can decide when it's recoverable and when it isn't.